lib/Data/Filesystem.php

```diff
@@ -409,38 +409,47 @@ function ($filename) {
      */
     public function getAllPastes()
     {
-        $pastes  = array();
-        $subdirs = scandir($this->_path);
-        if ($subdirs === false) {
-            dieerr('Unable to list directory ' . $this->_path);
-        }
-        $subdirs = preg_grep('/^[^.].$/', $subdirs);
-
-        foreach ($subdirs as $subdir) {
-            $subpath = $this->_path . DIRECTORY_SEPARATOR . $subdir;
-
-            $subsubdirs = scandir($subpath);
-            if ($subsubdirs === false) {
-                dieerr('Unable to list directory ' . $subpath);
-            }
-            $subsubdirs = preg_grep('/^[^.].$/', $subsubdirs);
-            foreach ($subsubdirs as $subsubdir) {
-                $subsubpath = $subpath . DIRECTORY_SEPARATOR . $subsubdir;
+        $pastes     = array();
+        $firstLevel = array_filter(
+            scandir($this->_path),
+            'PrivateBin\Data\Filesystem::_isFirstLevelDir'
+        );
+        if (count($firstLevel) > 0) {
+            foreach ($firstLevel as $firstKey) {
+                $secondLevel = array_filter(
+                    scandir($this->_path . DIRECTORY_SEPARATOR . $firstKey),
+                    'PrivateBin\Data\Filesystem::_isSecondLevelDir'
+                );
 
-                $files = scandir($subsubpath);
-                if ($files === false) {
-                    dieerr('Unable to list directory ' . $subsubpath);
+                // skip this folder
+                if (count($secondLevel) == 0) {
+                    continue;
                 }
-                $files = preg_grep('/\.php$/', $files);
 
-                foreach ($files as $file) {
-                    if (substr($file, 0, 4) === $subdir . $subsubdir) {
-                        $pastes[] = substr($file, 0, strlen($file) - 4);
+                foreach ($secondLevel as $secondKey) {
+                    $path = $this->_path . DIRECTORY_SEPARATOR . $firstKey .
+                        DIRECTORY_SEPARATOR . $secondKey;
+                    if (!is_dir($path)) {
+                        continue;
                     }
+                    $thirdLevel = array_filter(
+                        array_map(
+                            function ($filename) {
```

Comment:
Is that an inline function, uff? Can we at least give that a name and make it a non-anonymous function?

It seems some filename splitting? That could be a helper and also, does it assume a file extension always consists of 4 characters? (If so, that is obviously wrong, we have `.doc` and `.docx` e.g. Also I **hope** PHP has some library or so for doing that…)

So actionable again:
* Split that into own file/class/helper even. It's basic FS functionality. Clean-code wise.
* Maybe we can find a better library that does this task for us? 
* A short search also showed a better way using `explode` (looks better for me without knowing the details here)
  https://stackoverflow.com/a/37856144/5008962
* We can then add a proper parametrized test only for that small unit of splitting file names…

Reply:
It's used inside the filter which only returns valid paste IDs. In order to reuse that validation function, we need to remove the suffixes from the paste files - if they then aren't valid IDs, they simply are filtered out. Yes, this is all designed for the case that an admin ends up having garbage files inside the data directory for some reason.

Ideally we can use an operation that makes use of iterators, so we can stream through the scanned files as they are found. Maybe I can rewrite it using the dir class and it's read. Or we could try and use glob, as we know the exact structure of the directories and files: ab/cd/abcdef0123456789.php, then return the sub-string at the expected offsets.

Reply:
New solution doesn't need the filters anymore.